lib/services/service_brokers/v2/catalog.rb

```diff
@@ -122,7 +122,7 @@ def taken_names(new_services_names)
     end
 
     def can_delete_service?(service)
-      service.service_plans_dataset.map(&:service_instances_dataset).map(&:count).all?(0)
+      service.service_plans_dataset.all? { |plan| plan.service_instances_dataset.empty? }
```

Comment:
I think this could also be combined into a single SELECT statement with JOINs.

Did you check the actual SQL statements being executed?